core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableJoinRule.java

```diff
@@ -71,26 +72,34 @@
         EnumerableRules.LOGGER.debug(e.toString());
         return null;
       }
-    }
-    RelNode newRel;
-    try {
-      newRel = EnumerableJoin.create(
+    } else if (info.isEqui() && join.isNonCorrelateSemiJoin()) {
+      // This is a semi-join.
+      return EnumerableNestedLoopJoin.create(
```

Comment:
I think this rule should not generate EnumerableNestedLoopJoin. In my opinion it should be created only from LogicalCorrelate, i.e. something like:
a) LogicalJoin == [EnumerableJoinRule] ==> EnumerableJoin (now EnumerableHashJoin) / EnumerableThetaJoin
b) LogicalJoin == [JoinToCorrelateRule] ==> LogicalCorrelate == [EnumerableCorrelateRule] ==> EnumerableCorrelate (now EnumerableNestedLoop)

Reply:
This question is gone if we move semiJoin implementation to EnumerableHashJoin.